### ISSUE TITLE
分割交流

### ISSUE THREAD

**Issue by JWSunny** [niecongchong/DANet-keras#3]
你好，目前好像遇到相似的问题，在unet-resnet等模型上效果都是不好，val_acc上身一点就不再变化，方便交流下有尝试其他网络效果不错的，除了fcn8

**Comment by jayboxyz** [comment]
> 你好，目前好像遇到相似的问题，在unet-resnet等模型上效果都是不好，val_acc上身一点就不再变化，方便交流下有尝试其他网络效果不错的，除了fcn8

你还跑了哪些分割网络

**Comment by niecongchong** [comment]
建议跑一下作者原来的pytorch代码，之前用keras写的，自己搭的整个代码结构，可能不太好用。推荐你们用https://github.com/hszhao/semseg中的代码结构，搭配https://github.com/junfu1115/DANet中提供的DANet的模型代码，效果不错。

**Comment by jayboxyz** [comment]
> 建议跑一下作者原来的pytorch代码，之前用keras写的，自己搭的整个代码结构，可能不太好用。推荐你们用[https://github.com/hszhao/semseg中的代码结构，搭配https://github.com/junfu1115/DANet中提供的DANet的模型代码，效果不错。](https://github.com/hszhao/semseg%E4%B8%AD%E7%9A%84%E4%BB%A3%E7%A0%81%E7%BB%93%E6%9E%84%EF%BC%8C%E6%90%AD%E9%85%8Dhttps://github.com/junfu1115/DANet%E4%B8%AD%E6%8F%90%E4%BE%9B%E7%9A%84DANet%E7%9A%84%E6%A8%A1%E5%9E%8B%E4%BB%A3%E7%A0%81%EF%BC%8C%E6%95%88%E6%9E%9C%E4%B8%8D%E9%94%99%E3%80%82)

看了您的代码，请教下 danet 里的 attention 模块是可以插入任意 feature map 中是吗

![image](https://user-images.githubusercontent.com/25930007/70195319-4e772b80-1740-11ea-9175-c0b95975c556.png)

**Comment by niecongchong** [comment]
可以的，但是要考虑特征图尺寸问题，要不然算起来可能会很慢，

**Comment by swjtulinxi** [comment]
博主你好，你尝试过上面你说的那个源码吗，请问效果怎样，miou能上80吗，我正准备研究一下，如果效果一般我就不准备尝试了

**Comment by xiongyifeng007** [comment]
大神好，请问你这个代码怎么运行呀

**Comment by niecongchong** [comment]
@swjtulinxi 尝试过，在与danet原文配置相同的情况下复现，结果甚至比nyuanwen结果好，至于80，准确的说，不加多尺度测试到不了，原文也没到80，加了多尺度测试差不多

**Comment by niecongchong** [comment]
@xiongyifeng007 上面的代码很早前写的了，建议跑一下作者原来的pytorch代码，之前用keras写的，自己搭的整个代码结构，可能不太好用。推荐你们用https://github.com/hszhao/semseg中的代码结构，搭配https://github.com/junfu1115/DANet中提供的DANet的模型代码，效果不错。

**Comment by jayboxyz** [comment]
@niecongchong 大佬，我看你好像是做遥感相关的？想问下，你有把双重注意力用在遥感影像上实践吗，效果怎样？ ps 我目前在做遥感影像的分割~

**Comment by xiongyifeng007** [comment]
您好我是小白，我现在只想先跑通一个tensorflow的语义分割框架，任意一个网络模型就行，GitHub上面找了好多都不行 。请问你知道哪个能跑通吗，跪谢&nbsp; 老师把我放养了让我自己弄
 



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Nick Nie"<notifications@github.com&gt;; 
发送时间: 2019年12月25日(星期三) 下午3:42
收件人: "niecongchong/DANet-keras"<DANet-keras@noreply.github.com&gt;; 
抄送: "(8T13F16) 活 动 中 心"<214127235@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [niecongchong/DANet-keras] 分割交流 (#3)




@xiongyifeng007 上面的代码很早前写的了，建议跑一下作者原来的pytorch代码，之前用keras写的，自己搭的整个代码结构，可能不太好用。推荐你们用https://github.com/hszhao/semseg中的代码结构，搭配https://github.com/junfu1115/DANet中提供的DANet的模型代码，效果不错。

**Comment by niecongchong** [comment]
@strivebo 我也是渣渣一枚。我做的确实是遥感。关于danet，感觉发发论文还行，实际表现也还可以，不过17年的pspnet就可以和danet精度相当。

**Comment by niecongchong** [comment]
@xiongyifeng007 tensorflow说实话没搞过，github应该很多，你可以先找一找unet，pspnet的tf浮现代码。

**Comment by xiongyifeng007** [comment]
大哥 入门的话是PyTorch好还是tensorflow合适呢 我看网上都没有语义分割代码的实战教程
 



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Nick Nie"<notifications@github.com&gt;; 
发送时间: 2019年12月25日(星期三) 下午4:01
收件人: "niecongchong/DANet-keras"<DANet-keras@noreply.github.com&gt;; 
抄送: "(8T13F16) 活 动 中 心"<214127235@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [niecongchong/DANet-keras] 分割交流 (#3)




@xiongyifeng007 tensorflow说实话没搞过，github应该很多，你可以先找一找unet，pspnet的tf浮现代码。

**Comment by niecongchong** [comment]
入门肯定还是pytorch合适一些，关于pytorch入门，你只需要掌握基本的框架知识就行，边用边查。然后你找一个比较好的别人写过的分割代码，里面关于多卡并行，数据加载等等都有了，然后花点时间看懂，以后就在那个上面修改就行，可以参考pspnet作者的代码：https://github.com/hszhao/semseg

**Comment by jayboxyz** [comment]
> @strivebo 我也是渣渣一枚。我做的确实是遥感。关于danet，感觉发发论文还行，实际表现也还可以，不过17年的pspnet就可以和danet精度相当。

@niecongchong  我记得之前我应该是用了你写的 danet，好像 acc 上不去，但跑 pspnet 正常，acc 能 0.9 多，另外，fcn8下也还可以。unet、segnet 不太行，存在一些问题...好奇怪...  不知道是数据集的问题还是代码问题...

对了，话说您都是玩的哪个数据集哈？isprs 吗？

**Comment by niecongchong** [comment]
之前写的danet的attention模块确实有问题，一直也没时间回去改一改。

**Comment by niecongchong** [comment]
@strivebo 我之前一直用的isprs，今年下半年跑的cityscapes，

**Comment by swjtulinxi** [comment]
但我根据论文来读你的代码，感觉是这个流程啊------------------ 原始邮件 ------------------
发件人: "Nick Nie"<notifications@github.com>
发送时间: 2019年12月25日(星期三) 下午4:53
收件人: "niecongchong/DANet-keras"<DANet-keras@noreply.github.com>;
抄送: "swjtulinxi"<1335452106@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [niecongchong/DANet-keras] 分割交流 (#3)



之前写的danet的attention模块确实有问题，一直也没时间回去改一改。
 
&mdash;

**Comment by niecongchong** [comment]
你对比着看一下pytorch版本的代码，还是有些不一样的

**Comment by YLONl** [comment]
@niecongchong .作者你好，
①我最近也在尝试复现pytorch的danet，有些是可以进行训练，但当我把数据集换成医学灰度图像的时候，则会出现很多问题，我尝试了各种都不行。包括改变输入网络参数的channels，图像灰度图的改变，不知道你有没有尝试过灰度图的数据集的分割，我以前用keras的u-net进行分割的时候是可以的。复现的是这一个网络（https://github.com/Tramac/awesome-semantic-segmentation-pytorch里面有danet），不知道你复现过没有。
②还有一个pytorch的包导入问题No module named 'enclib_gpu' 或'enclib_cpu'（https://github.com/junfu1115/DANet），网上的解决办法都尝试了，都没有成功，不知你遇到过没有。

**Comment by YLONl** [comment]
> @xiongyifeng007 上面的代码很早前写的了，建议跑一下作者原来的pytorch代码，之前用keras写的，自己搭的整个代码结构，可能不太好用。推荐你们用[https://github.com/hszhao/semseg中的代码结构，搭配https://github.com/junfu1115/DANet中提供的DANet的模型代码，效果不错。](https://github.com/hszhao/semseg%E4%B8%AD%E7%9A%84%E4%BB%A3%E7%A0%81%E7%BB%93%E6%9E%84%EF%BC%8C%E6%90%AD%E9%85%8Dhttps://github.com/junfu1115/DANet%E4%B8%AD%E6%8F%90%E4%BE%9B%E7%9A%84DANet%E7%9A%84%E6%A8%A1%E5%9E%8B%E4%BB%A3%E7%A0%81%EF%BC%8C%E6%95%88%E6%9E%9C%E4%B8%8D%E9%94%99%E3%80%82)

作者，你这样进行搭配是因为junfu1115/DANet的网络是没有跑成功吗?

**Comment by renyibinouc** [comment]
###CAM
.....
softmax_aTa = Activation('softmax')(aTa)
softmax_aTa = tf.transpose(softmax_aTa, (0, 2, 1)) ##增加这一句
aaTa = K.batch_dot(vec_a, softmax_aTa)
......

CAM中是否应该增加一句代码，把softmax_aTa 转置，softmax_aTa 的每一行的和是1，转置后变成每一列的和是1，这样vec_a * softmax_aTa 才有意义

**Comment by swjtulinxi** [comment]
你可以看下源码。其实这个注意力nonlocal模块也就是炒概念，精度的化也只能和deeplab的aspp模块相当

**Comment by CXxian708** [comment]
请问，attention模块中问题在那个地方，我运行的时候会出问题，但是还不知具体怎么改。请问有没有改好的code可以分享或者推荐一下。
